src/pytest_cov/plugin.py

```diff
@@ -180,30 +184,55 @@ def pytest_testnodedown(self, node, error):
         self.cov_controller.testnodedown(node, error)
     pytest_testnodedown.optionalhook = True
 
-    def pytest_sessionfinish(self, session, exitstatus):
-        """Delegate to our implementation."""
-        self.failed = exitstatus != 0
+    def _should_report(self):
+        return not (self.failed and self.options.no_cov_on_fail)
+
+    def _failed_cov_total(self):
+        cov_fail_under = self.options.cov_fail_under
+        return cov_fail_under is not None and self.cov_total < cov_fail_under
+
+    # we need to wrap pytest_runtestloop. by the time pytest_sessionfinish
+    # runs, it's too late to set testsfailed
+    @compat.hookwrapper
+    def pytest_runtestloop(self, session):
+        outcome = yield
+
+        compat_session = compat.SessionWrapper(session)
+
+        self.failed = bool(compat_session.testsfailed)
         if self.cov_controller is not None:
             self.cov_controller.finish()
 
-    def pytest_terminal_summary(self, terminalreporter):
-        """Delegate to our implementation."""
-        if self.cov_controller is None:
-            return
-        if not (self.failed and self.options.no_cov_on_fail):
+        if self._should_report():
             try:
-                total = self.cov_controller.summary(terminalreporter.writer)
+                self.cov_total = self.cov_controller.summary(self.cov_report)
             except CoverageException as exc:
-                terminalreporter.writer.write('Failed to generate report: %s\n' % exc)
-                total = 0
-            assert total is not None, 'Test coverage should never be `None`'
-            cov_fail_under = self.options.cov_fail_under
-            if cov_fail_under is not None and total < cov_fail_under:
                 raise pytest.UsageError(
-                    'Required test coverage of %d%% not '
-                    'reached. Total coverage: %.2f%%\n'
-                    % (self.options.cov_fail_under, total)
+                    'Failed to generate report: %s\n' % exc
                 )
+            assert self.cov_total is not None, 'Test coverage should never be `None`'
+            if self._failed_cov_total():
+                # make sure we get the EXIT_TESTSFAILED exit code
+                compat_session.testsfailed += 1
+
+    def pytest_terminal_summary(self, terminalreporter):
+        if self.cov_controller is None:
+            return
+
+        if self.cov_total is None:
+            # we shouldn't report, or report generation failed (error raised above)
```

Comment:
We should still have the old assertion no? Or `pytest_terminal_summary` is called regardless of `pytest_runtestloop` raising an error? When does that happen?